### PR TITLE
Feature/wasm pack

### PR DESCRIPTION
**Pull request by maebli**
I've added wasm package.

I am looking into using serde crate. Need to know if this is correct. 

Serde should be avilable in cli and wasm but not as an option in the lib

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -28,15 +28,19 @@ bindgen = "0.69.4"
...
 lto = true        # Enable Link Time Optimization
 codegen-units = 1 # Reduce codegen units to improve optimizations
 
 [dependencies]
+serde = { version = "1.0", features = ["derive"], optional = true }
 bitflags = "2.4.2"
-arrayvec = "0.7.4"
+arrayvec = { version = "0.7.4", features = [
+    "serde",
+], optional = true, default-features = false }
```
Comment (SarthakSingh31) [inline Cargo.toml:43]: Is arrayvec really optional? You seem to use it without a feature gate in the code.
I think you just want
`arrayvec = { version = "0.7.4", default-features = false }`

Comment (SarthakSingh31) [reply Cargo.toml:43]: The `arrayvec/serde`  feature flag will enable the serde feature on arrayvec when necessary

File: src/lib.rs [hunk 2]
```diff
@@ -58,9 +58,16 @@ pub mod frames;
 pub mod user_data;
 
 #[derive(Debug)]
+#[cfg_attr(feature = "serde_support", derive(serde::Serialize))]
+#[cfg_attr(
+    feature = "serde_support",
+    derive(serde::Deserialize),
+    serde(bound(deserialize = "'de: 'a"))
+)]
```
Comment (SarthakSingh31) [inline src/lib.rs:66]: These can probably be the same cfg_attr. You don't need a separate one for each derive.
```rust
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize),
    serde(bound(deserialize = "'de: 'a"))
)]
```

File: src/user_data/data_information.rs [hunk 3]
```diff
@@ -3,9 +3,14 @@ use super::variable_user_data::DataRecordError;
...
 
 const MAX_DIFE_RECORDS: usize = 10;
+#[cfg_attr(
+    feature = "serde_support",
+    derive(serde::Serialize, serde::Deserialize)
+)]
 #[derive(Debug, PartialEq)]
 pub struct DataInformationBlock {
     pub data_information_field: DataInformationField,
+    #[serde(skip_serializing, skip_deserializing)]
```
Comment (SarthakSingh31) [inline src/user_data/data_information.rs:13]: this should also be in a cfg_attr

File: src/user_data/mod.rs [hunk 4]
```diff
@@ -61,6 +65,26 @@ impl Default for DataRecords {
...
+impl<'de> serde::Deserialize<'de> for StatusField {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let bits = u8::deserialize(deserializer)?;
+        StatusField::from_bits(bits)
+            .ok_or_else(|| serde::de::Error::custom("Invalid bits for StatusField"))
+    }
+}
```
Comment (SarthakSingh31) [inline src/user_data/mod.rs:87]: You can derive these with cfg_attr in the bitflags macro.

File: wasm/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+mod utils;
+use m_bus_parser::{self, MbusData};
+
+#[cfg(feature = "wee_alloc")]
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
```
Comment (SarthakSingh31) [inline wasm/src/lib.rs:6]: Don't use weealloc, its very broken. See issues:
- https://github.com/rustwasm/wee_alloc/issues/109
- https://github.com/rustwasm/wee_alloc/issues/106
- https://github.com/rustwasm/wee_alloc/issues/105

File: src/user_data/mod.rs (hunk 4 above)
Comment (maebli) [reply src/user_data/mod.rs:87]: somehow this was not working

Comment (maebli) [reply src/user_data/mod.rs:87]: Screenshot 2024-06-12 at 21 44 00

File: src/lib.rs (hunk 2 above)
Comment (maebli) [reply src/lib.rs:66]: this was not possible

File: src/user_data/mod.rs (hunk 4 above)
Comment (SarthakSingh31) [reply src/user_data/mod.rs:84]: what error does it show? I am doing this cargo check and it shows no error.

Comment (SarthakSingh31) [reply src/user_data/mod.rs:84]: oh, you also need to add bitflags/serde to the serde feature flag

Comment (SarthakSingh31) [reply src/user_data/mod.rs:84]: made a pr with this https://github.com/maebli/m-bus-parser/pull/11

**Pull request merged**